core/src/main/scala/kafka/zk/KafkaZkClient.scala

```diff
@@ -1631,10 +1637,37 @@ class KafkaZkClient private (zooKeeperClient: ZooKeeperClient, isSecure: Boolean
       throw KeeperException.create(code)
   }
 
+  private def isZKSessionIdDiffFromCurrentZKSessionId(): Boolean = {
+    zooKeeperClient.sessionId != currentZooKeeperSessionId
+  }
+
+  private def isZKSessionTheEphemeralOwner(ephemeralOwnerId: Long): Boolean = {
+    ephemeralOwnerId == currentZooKeeperSessionId
+  }
+
+  private[zk] def shouldReCreateEphemeralZNode(ephemeralOwnerId: Long): Boolean = {
+    isZKSessionTheEphemeralOwner(ephemeralOwnerId) && isZKSessionIdDiffFromCurrentZKSessionId()
+  }
+
+  private def updateCurrentZKSessionId(newSessionId: Long): Unit = {
+    currentZooKeeperSessionId = newSessionId
+  }
+
+  private def setCurrentZookeeperSessionId(): Unit = {
+    if ( currentZooKeeperSessionId == -1 ) {
+      currentZooKeeperSessionId = zooKeeperClient.sessionId
+    }
+  }
+
   private class CheckedEphemeral(path: String, data: Array[Byte]) extends Logging {
     def create(): Code = {
       val createRequest = CreateRequest(path, data, acls(path), CreateMode.EPHEMERAL)
       val createResponse = retryRequestUntilConnected(createRequest)
+      // At this point, we need to save a reference to the zookeeper session id, just for the first time.
+      // That is, the fist attempt to register the Broker, we save a reference of the used Zookeeper session id.
+      // This is done here since the Zookeeper session id may not be available at the Object creation time,
+      // hence, this is assuming the 'retryRequestUntilConnected' method got connected and a valid session id is present.
+      setCurrentZookeeperSessionId()
```

Comment:
Hmm, do we need both setCurrentZookeeperSessionId() and updateCurrentZKSessionId()?  To me, it seems that we can just initialize currentZooKeeperSessionId  to -1 and set currentZooKeeperSessionId 
 in checkedEphemeralCreate(path: String, data: Array[Byte]) if the return code is ok.